### ISSUE TITLE
Request: `clamp(x::AbstractFloat, ::Type{Integer})`

### ISSUE THREAD

**Issue by jebej** [JuliaLang/julia#40469]
I'd like to request a function to clamp floating point values and subsequently round to the nearest integer. Note that such a function does exist for integer - integer combinations, but fails with floating point inputs. 
```julia
julia> clamp(21, Int8)
21
julia> clamp(21.3, Int)
ERROR: MethodError: no method matching rem(::Float64, ::Type{Int64})
```

This has been discussed [on Discourse](https://discourse.julialang.org/t/clamp-floats-to-int/), but the solution offered fails with large inputs and wraps around:
```julia
julia> clamp_to(T, x) = unsafe_trunc(T, clamp(x, typemin(T), typemax(T)))
clamp_to (generic function with 1 method)

julia> clamp_to(Int, 1e33)
-9223372036854775808
```

A simple implementation also fails due to round tripping issues (in this example, I'm sure there are other issues):
```julia
julia> round_clamp(::Type{T}, x::S) where {T,S} = round(T, clamp(x, S(typemin(T)), S(typemax(T))))
round_clamp (generic function with 1 method)

julia> round_clamp(Int, 1e21)
ERROR: InexactError: trunc(Int64, 9.223372036854776e18)

julia> round(Int, Float64(typemax(Int))) # caused by roundtripping failure
ERROR: InexactError: trunc(Int64, 9.223372036854776e18)
```

**Comment by tomyun** [comment]
I also had a similar request on `trunc/round/ceil/floor(T, x)` for Real `T`: #37984

**Comment by regier21** [comment]
Hmm so looking into why the `clamp_to` solution didn't work, I noticed that 
```julia 
clamp(1e33, typemin(Int), typemax(Int)) > typemax(Int)
``` 
This is caused by rounding to the nearest floating point number. While I understand why this occurs, it feels to me like it violates the contract for `clamp`, and it's what leads to the number becoming negative when converted.

I was hoping we could manually round the min and max sent to clamp down to the nearest floating point below the typemin and typemax. But this doesn't work either! I reduced the number of significant digits for testing, but regardless this happens:
```julia
julia> floor(typemax(Int), sigdigits=3, base=2) > typemax(Int)
true
```
It looks to me that we lack the ability to prevent the result of the clamp from being outside of the clamp unless we could somehow control the rounding in the conversion process.

**Comment by vtjnash** [comment]
Implementing #37984 then using that here seems like a good idea. I'll comment there on a possible implementation

**Comment by regier21** [comment]
Actually, it turns out Julia already has a function for this: `prevfloat`. So this should work:
```julia
julia> function clamp_to(T, x)
           clamped = clamp(x, typemin(T), typemax(T))
           if clamped > typemax(T)
               clamped = prevfloat(clamped)
           elseif clamped < typemin(T)
               clamped = nextfloat(clamped)
           end
           return unsafe_trunc(T, clamped)
       end
clamp_to (generic function with 1 method)

julia> clamp_to(Int, 1e33)
9223372036854774784
```

**Comment by regier21** [comment]
> Implementing #37984 then using that here seems like a good idea. I'll comment there on a possible implementation

Ha I was typing out my solution when you posted that. Looks like we came up with pretty similar solutions, although I think I prefer modifying the `floor` function directly like you did.

**Comment by vtjnash** [comment]
I'm not certain if this is intended or not though, since it seems `floor` should represent the closest integer, not the closest representable value, so perhaps it is `clamp` that would be better to fix for both cases.
```
julia> convert(Int, convert(Float64, typemax(Int)))
ERROR: InexactError: Int64(9.223372036854776e18)
...
```

**Comment by regier21** [comment]
When the named arguments `sigdigits` or `digits` are passed into `floor`, then the result is the nearest float, not the nearest integer. I would hope that `floor(x, sigdigits=8*sizeof(x), base=2) <= x`, but that isn't true right now because of this weird rounding behavior for very large `x`.

**Comment by regier21** [comment]
Shoot I just realized my solution doesn't correctly return `typemax(Int)` because it reduces the float before truncating. The code should return `typemax(x)` instead of calling `prevfloat` (and likewise for the other case).

**Comment by jebej** [comment]
Yes, IMO, a working implementation would have the following behavior, though maybe even that one would have issues with the comparisons:
```julia
function clamp(x::AbstractFloat, ::Type{T}) where T<:Integer
    if x > typemax(T)
        return typemax(T)
    elseif x < typemin(T)
        return typemin(T)
    else
        return round(T, x)
    end
end
```

**Comment by jebej** [comment]
Yes, the above would fail for some integers near `typemax(Int)`:
```julia
julia> 9223372036854775600 < typemax(Int)
true

julia> 9223372036854775600.0 < typemax(Int)
false
```

**Comment by goretkin** [comment]
This may be too off-topic, but I wonder if this is an opportunity to introduce a `project` (the verb corresponding to "projection", though I hope a less confusing name exists) that takes 1. a domain, 2. an object, 3. optionally a "metric".

The domain could be a type, but also something like `Set([1,2,3])` or `10:40` or an interval (outside `Base`) `3.1..4.9`.
The result would be an element "in" the domain.

e.g.  rounding:
`project(Int64, 3.1, RoundNearest)`

e.g. clamping:
`project(4...5, 3.1)`

It may even be worth introducing a wrapper type so that we could say `in` the domain vs "in" the domain: `project(d, ...) ∈ d`.

```julia
julia> struct TypesAreSetsOfValues{T} end

julia> Base.in(x, ::TypesAreSetsOfValues{T}) where{T} = typeof(x) <: T

julia> 3 ∈ TypesAreSetsOfValues{Int64}()
true

julia> 3.0 ∈ TypesAreSetsOfValues{Int64}()
false
```

Obviously that's getting verbose, so an interface could automatically wrap `<:Number` in something like that. Also, `Base` does not have a built-in interval type (except for e.g. `Int64`), so it would not be as easy to use as `clamp(val, low, high)`.

**Comment by vtjnash** [comment]
If you look very closely and carefully, we can see that it actually is greater:
```
julia> @printf "%.0f" 9223372036854775600.0
9223372036854775808
julia> typemax(Int)
9223372036854775807
```

**Comment by bjarthur** [comment]
i would find `clamp(x::AbstractFloat, ::Type{Integer})` useful too
